### ISSUE TITLE
Quick fix for Debug.LogFormat

### ISSUE THREAD

**Issue by Chippit** [JetBrains/resharper-unity#606]
Rider offers a quick fix to transform concatenated strings into a String.Format invocation instead (or string interpolation if available in language level)

It'd be great if it's possible that, if concatenation appears inside a Debug.Log (or Warning/Error variants), that it offers a suggestion to use Debug.LogFormat instead.

**Comment by van800** [comment]
@Chippit Do you mean for
Debug.LogFormat("last {0}", 1);
proposing 
Debug.LogFormat($"last {1}");
?

**Comment by StephenHodgson** [comment]
@van800 depends on if they've got 4.x scripting runtime enabled, or else they won't get string interpolation.

**Comment by StephenHodgson** [comment]
I think @Chippit is trying to describe:
`Debug.Log("My Error: " + value);` vs `Debug.LogFormat("My Error: {0}", value);`

**Comment by van800** [comment]
Rider is smart enough to know the context when string interpolation is applicable. The problem here is that in the proposed example there is no reasonable way to analyze that string and arguments would be concatenated.

**Comment by Chippit** [comment]
Apologies, I may not have worded my initial suggestion well. @StephenHodgson has it right.

Rider seems to have behaviour similar to this already for calls to such things as Console.WriteLine (or indeed custom methods decorated with StringFormatMethodAttribute), presumably currently only working for overloads.

But for instance, in cases where string interpolation is not available, the quick fix "To String.Format invocation" for, say, `Console.WriteLine("Text " + value);` transforms it into `Console.WriteLine("Text {0}", value);`

The same thing on `Debug.Log("Test " + value);` transforms it into a (not unreasonable) `Debug.Log(string.Format("Test {0}", value));`. But in this case, `Debug.LogFormat("Test {0}", value);` would have been preferable, since Unity does not offer a formatting overload for Debug.Log (probably they should, but this'd mean changing their current Log API, and they'd be unsurprisingly hesitant to do so!).

**Comment by StephenHodgson** [comment]
They actually do provide an overload.

`Debug.LogFormat("this {0}", thing);`

**Comment by StephenHodgson** [comment]
Seems redundant though if ppl are using 4.x language features like string interpolation.

**Comment by citizenmatt** [comment]
So, to sum up, the request is to recognise when `Debug.Log("something" + value)` is being used and suggest to use `Debug.LogFormat("something {0}", value)` instead. This should be fairly easy - ReSharper already has context actions to do similar things with `String.Format` and `Console.WriteLine`, and I think we can reuse some of the logic here.

However, I'm not sure if this should be a highlight (warning/suggestion) or just a context action, as the `LogFormat` version doesn't offer much benefit other than being cleaner. Both `Debug.Log` and `Debug.LogFormat` will allocate - one for string concatenation and the other for `params object[]`.

(It would be much better if `Debug.Log`/`Debug.LogFormat` was a conditional method. Or offered multiple overloads with generic methods. The .net framework has multiple overloads of `string.Format` to avoid allocating the object array, but makes all of the parameters `object`, so you still get boxing)

**Comment by Chippit** [comment]
Agreed, I don't believe it should be a highlight either, just a potential context action. It simply feels like a nice convenience that'll make Rider feel smarter and more aware of Unity capabilities. It might also be a little less important now that Unity properly supports 4.x features, but that's also only available for users on the latest versions of Unity - I expect many are still using pre-2018 versions where this suggestion would be most relevant.

I also agree about Debug.Log methods being conditional. For all of our projects, we've always used wrappers that are so that we might strip those calls to eliminate the potential allocations. However, that's well beyond the scope of Rider's capabilities anyway :)
